### Сравнение предложений

In [3]:
import re
import numpy as np
from collections import Counter

In [10]:
!head sentences.txt

In comparison to dogs, cats have not undergone major changes during the domestication process.
As cat simply catenates streams of bytes, it can be also used to concatenate binary files, where it will just concatenate sequence of bytes.
A common interactive use of cat for a single file is to output the content of a file to standard output.
Cats can hear sounds too faint or too high in frequency for human ears, such as those made by mice and other small animals.
In one, people deliberately tamed cats in a process of artificial selection, as they were useful predators of vermin.
The domesticated cat and its closest wild ancestor are both diploid organisms that possess 38 chromosomes and roughly 20,000 genes.
Domestic cats are similar in size to the other members of the genus Felis, typically weighing between 4 and 5 kg (8.8 and 11.0 lb).
However, if the output is piped or redirected, cat is unnecessary.
cat with one named file is safer where human error is a concern - one wrong use of the

In [11]:
with open('sentences.txt') as file_:
    sentences = file_.read()

Каждая строка в файле соответствует одному предложению. Считайте их, приведите каждую к нижнему регистру с помощью строковой функции lower().

In [12]:
sentences = sentences.lower()

Произведите токенизацию, то есть разбиение текстов на слова. Для этого можно воспользоваться регулярным выражением, которое считает разделителем любой символ, не являющийся буквой: re.split('[^a-z]', t). Не забудьте удалить пустые слова после разделения.

In [13]:
sentences = [s for s in re.split('\.\n', sentences) if s != '']
# Получен список предложений

In [14]:
len(sentences)

22

In [15]:
tokenized_sent = [re.split('[^a-z]', s) for s in sentences]
# предложения токенизированы

In [16]:
tokenized_sentences = []
for tok in tokenized_sent:
    tokenized_sentences.append([i for i in tok if i != ''])
# удалены пустые токены

In [ ]:
# tokenized_sentences = []
# for sentence in sentences:
#     tokens = re.split('[^a-z]', sentence)
#     tokenized_sentences.append(filter(lambda t : t != "", tokens))

Составьте список всех слов, встречающихся в предложениях. Сопоставьте каждому слову индекс от нуля до (d - 1), где d — число различных слов в предложениях. Для этого удобно воспользоваться структурой dict.

In [24]:
words = [word for sent in tokenized_sentences for word in sent]
uniq_words = set(words)
# составлен список уникальных слов

In [25]:
len(uniq_words)

254

In [26]:
word_dict = dict()

for numb, i in enumerate(uniq_words):
    word_dict[i] = numb

In [ ]:
# word_dict = {}

# word_id = 0
# for word in words:
#     if word not in word_dict:
#         word_dict[word] = word_id
#         word_id += 1

In [29]:
word_dict

{'a': 232,
 'according': 98,
 'adjacent': 211,
 'allow': 248,
 'allows': 97,
 'also': 208,
 'an': 188,
 'ancestor': 215,
 'and': 177,
 'animals': 164,
 'any': 196,
 'app': 114,
 'apple': 113,
 'are': 58,
 'arguments': 210,
 'artificial': 68,
 'as': 190,
 'available': 51,
 'based': 88,
 'basic': 138,
 'be': 79,
 'between': 72,
 'binary': 199,
 'both': 91,
 'by': 90,
 'bytes': 42,
 'can': 172,
 'cat': 170,
 'catenates': 106,
 'cats': 136,
 'changes': 107,
 'chromosomes': 218,
 'clear': 239,
 'closest': 219,
 'command': 246,
 'commands': 92,
 'common': 77,
 'community': 115,
 'comparison': 230,
 'computers': 179,
 'concatenate': 78,
 'concern': 62,
 'connected': 227,
 'contains': 148,
 'content': 69,
 'count': 162,
 'create': 130,
 'default': 167,
 'delete': 125,
 'deliberately': 25,
 'developed': 5,
 'diploid': 85,
 'disk': 15,
 'displays': 0,
 'dogs': 141,
 'domestic': 8,
 'domesticated': 195,
 'domestication': 197,
 'download': 60,
 'drive': 22,
 'during': 37,
 'ears': 50,
 'editions':

In [28]:
len(word_dict)

254

In [30]:
temp_sentences = []
for tok in tokenized_sentences:
    temp_sentences.append([word_dict[word] for word in tok])
# шифруем предложения нашими числовыми значениями

In [34]:
temp_sentences[:1]

[[193, 230, 23, 141, 136, 180, 36, 124, 110, 107, 37, 251, 197, 182]]

Создайте матрицу размера n X d, где n — число предложений. Заполните ее: элемент с индексом (i, j) в этой матрице должен быть равен количеству вхождений j-го слова в i-е предложение. У вас должна получиться матрица размера 22 * 254.

In [35]:
sent_matrix = np.zeros((len(sentences), len(uniq_words)))

In [36]:
sent_matrix.shape

(22, 254)

In [37]:
for num_sentence, sentence in enumerate(temp_sentences):
    counter_dict = Counter(sentence)
    for word in sentence:
        value = counter_dict[word]
        sent_matrix[num_sentence, word] = value

In [38]:
sent_matrix[21][np.newaxis, : ].shape

(1, 254)

Найдите косинусное расстояние от предложения в самой первой строке (In comparison to dogs, cats have not undergone...) до всех остальных с помощью функции scipy.spatial.distance.cosine. Какие номера у двух предложений, ближайших к нему по этому расстоянию (строки нумеруются с нуля)? Эти два числа и будут ответами на задание.

In [49]:
from scipy.spatial.distance import cosine
# косинусное расстояние. Чем меньше, тем сильнее схожи 2 предложения

In [50]:
cos_vectors = []

for i in sent_matrix:
    cos_vectors.append(cosine(sent_matrix[0], i))
#     print cosine(sent_matrix[0], i)

In [51]:
cos_vectors

[0.0,
 0.95275444087384664,
 0.86447381456421235,
 0.89517151632780823,
 0.77708871496985887,
 0.94023856953328033,
 0.7327387580875756,
 0.92587506833388988,
 0.88427248752843102,
 0.90550888174769317,
 0.83281653622739416,
 0.88047713906656067,
 0.83964325485254543,
 0.87035925528956715,
 0.87401184233025764,
 0.94427217874246472,
 0.84063618542208085,
 0.95664450152379399,
 0.94427217874246472,
 0.88854435748492944,
 0.84275727449171223,
 0.82503644694405864]

In [54]:
sorted_vec = sorted(cos_vectors)

In [55]:
sorted_vec[1:3]

[0.7327387580875756, 0.77708871496985887]

In [56]:
close_sent = []
c = 0
for i in cos_vectors:
    c += 1
    if i in sorted_vec[1:3]:
        close_sent.append(c-1)

In [52]:
# np.argsort(cos_vectors)[:3]

In [53]:
# from operator import itemgetter

# sorted(enumerate(cos_vectors), key=itemgetter(1))[:3]

In [57]:
close_sent

[4, 6]

In [58]:
print tokenized_sentences[0]
for i in close_sent:
    print tokenized_sentences[i]

['in', 'comparison', 'to', 'dogs', 'cats', 'have', 'not', 'undergone', 'major', 'changes', 'during', 'the', 'domestication', 'process']
['in', 'one', 'people', 'deliberately', 'tamed', 'cats', 'in', 'a', 'process', 'of', 'artificial', 'selection', 'as', 'they', 'were', 'useful', 'predators', 'of', 'vermin']
['domestic', 'cats', 'are', 'similar', 'in', 'size', 'to', 'the', 'other', 'members', 'of', 'the', 'genus', 'felis', 'typically', 'weighing', 'between', 'and', 'kg', 'and', 'lb']


Совпадают ли ближайшие два предложения по тематике с первым? Совпадают ли тематики у следующих по близости предложений?

> ближайшие 2 предложения совпадают по тематике с первым (тема: коты), но следующие по близости уже нет (тема: мак ос)

### cosine_similarity

In [59]:
from sklearn.metrics.pairwise import cosine_similarity
true_vectors = []

for i in sent_matrix:
    true_vectors.append(float(cosine_similarity(sent_matrix[0][np.newaxis, : ], i[np.newaxis, : ])))

In [60]:
cosine_similarity(sent_matrix[0][np.newaxis, : ], i[np.newaxis, : ])

array([[ 0.17496355]])

In [64]:
close_true_sent = []
c = 0
for i in true_vectors:
    c += 1
    if i in sorted(true_vectors, reverse=True)[1:3]:
        close_true_sent.append(c-1)

In [65]:
print close_true_sent

print tokenized_sentences[0]
for i in close_true_sent:
    print tokenized_sentences[i]

[4, 6]
['in', 'comparison', 'to', 'dogs', 'cats', 'have', 'not', 'undergone', 'major', 'changes', 'during', 'the', 'domestication', 'process']
['in', 'one', 'people', 'deliberately', 'tamed', 'cats', 'in', 'a', 'process', 'of', 'artificial', 'selection', 'as', 'they', 'were', 'useful', 'predators', 'of', 'vermin']
['domestic', 'cats', 'are', 'similar', 'in', 'size', 'to', 'the', 'other', 'members', 'of', 'the', 'genus', 'felis', 'typically', 'weighing', 'between', 'and', 'kg', 'and', 'lb']
